File: src/utils/analyzer.ts
import { ScoredWallet } from './modelTrainer';
import { WalletFeatures } from './featureEngineer';

export interface AnalysisResult {
  score_distribution: {
    total_wallets: number;
    mean_score: number;
    median_score: number;
    std_score: number;
    min_score: number;
    max_score: number;
    bucket_distribution: Record<string, { count: number; percentage: number }>;
    risk_category_distribution: Record<string, { count: number; percentage: number }>;
  };
  behavioral_patterns: {
    high_score: any;
    medium_score: any;
    low_score: any;
  };
  risk_factors: {
    score_correlations: Record<string, number>;
    high_risk_characteristics: {
      count: number;
      avg_liquidations: number;
      avg_leverage: number;
      bot_like_ratio: number;
    };
  };
  comparative_analysis: {
    high_scorers: { count: number; score_range: string };
    low_scorers: { count: number; score_range: string };
    metric_differences: Record<string, {
      high_scorers_avg: number;
      low_scorers_avg: number;
      difference_ratio: number;
    }>;
  };
}

export function generateAnalysis(scores: ScoredWallet[], features: WalletFeatures[]): AnalysisResult {
  console.log('Generating comprehensive analysis...');
  
  const scoreValues = scores.map(s => s.credit_score);
  const totalWallets = scores.length;

  // Score distribution
  const scoreDistribution = {
    total_wallets: totalWallets,
    mean_score: mean(scoreValues),
    median_score: median(scoreValues),
    std_score: standardDeviation(scoreValues),
    min_score: Math.min(...scoreValues),
    max_score: Math.max(...scoreValues),
    bucket_distribution: calculateBucketDistribution(scoreValues),
    risk_category_distribution: calculateRiskCategoryDistribution(scores)
  };

  // Merge scores with features for analysis
  const mergedData = scores.map(score => {
    const walletFeatures = features.find(f => f.wallet_address === score.wallet_address);
    return { ...score, ...walletFeatures };
  });

  // Behavioral patterns
  const behavioralPatterns = analyzeBehavioralPatterns(mergedData);
  
  // Risk factors
  const riskFactors = analyzeRiskFactors(mergedData);
  
  // Comparative analysis
  const comparativeAnalysis = analyzeComparative(mergedData);

  return {
    score_distribution: scoreDistribution,
    behavioral_patterns: behavioralPatterns,
    risk_factors: riskFactors,
    comparative_analysis: comparativeAnalysis
  };
}

function calculateBucketDistribution(scores: number[]): Record<string, { count: number; percentage: number }> {
  const buckets: Record<string, { count: number; percentage: number }> = {};
  
  for (let i = 0; i < 10; i++) {
    const min = i * 100;
    const max = (i + 1) * 100 - 1;
    const key = `${min}-${max}`;
    const count = scores.filter(score => score >= min && score <= max).length;
    
    buckets[key] = {
      count,
      percentage: (count / scores.length) * 100
    };
  }
  
  return buckets;
}

function calculateRiskCategoryDistribution(scores: ScoredWallet[]): Record<string, { count: number; percentage: number }> {
  const categories: Record<string, number> = {};
  
  scores.forEach(score => {
    categories[score.risk_category] = (categories[score.risk_category] || 0) + 1;
  });
  
  const distribution: Record<string, { count: number; percentage: number }> = {};
  Object.entries(categories).forEach(([category, count]) => {
    distribution[category] = {
      count,
      percentage: (count / scores.length) * 100
    };
  });
  
  return distribution;
}

function analyzeBehavioralPatterns(mergedData: any[]) {
  const highScore = mergedData.filter(d => d.credit_score >= 700);
  const mediumScore = mergedData.filter(d => d.credit_score >= 400 && d.credit_score < 700);
  const lowScore = mergedData.filter(d => d.credit_score < 400);

  const analyzeGroup = (group: any[]) => {
    if (group.length === 0) return { count: 0 };
    
    return {
      count: group.length,
      total_transactions: { mean: mean(group.map(g => g.total_transactions || 0)) },
      account_age_days: { mean: mean(group.map(g => g.account_age_days || 0)) },
      repayment_ratio: { mean: mean(group.map(g => g.repayment_ratio || 0)) },
      leverage_ratio: { mean: mean(group.map(g => g.leverage_ratio || 0)) },
      asset_diversity_score: { mean: mean(group.map(g => g.asset_diversity_score || 0)) },
      liquidation_ratio: { mean: mean(group.map(g => g.liquidation_ratio || 0)) },
      transaction_complexity: { mean: mean(group.map(g => g.transaction_complexity || 0)) }
    };
  };

  return {
    high_score: analyzeGroup(highScore),
    medium_score: analyzeGroup(mediumScore),
    low_score: analyzeGroup(lowScore)
  };
}

function analyzeRiskFactors(mergedData: any[]) {
  // Calculate correlations with credit score
  const correlations: Record<string, number> = {};
  const riskFactors = [
    'liquidation_count', 'liquidation_ratio', 'leverage_ratio',
    'repayment_ratio', 'time_regularity_score', 'amount_uniformity_score'
  ];

  riskFactors.forEach(factor => {
    const scores = mergedData.map(d => d.credit_score);
    const factorValues = mergedData.map(d => d[factor] || 0);
    correlations[factor] = calculateCorrelation(scores, factorValues);
  });

  // High-risk characteristics
  const highRiskWallets = mergedData.filter(d => d.credit_score < 400);
  const highRiskChars = {
    count: highRiskWallets.length,
    avg_liquidations: mean(highRiskWallets.map(w => w.liquidation_count || 0)),
    avg_leverage: mean(highRiskWallets.map(w => w.leverage_ratio || 0)),
    bot_like_ratio: highRiskWallets.filter(w => (w.time_regularity_score || 0) > 2).length / highRiskWallets.length
  };

  return {
    score_correlations: correlations,
    high_risk_characteristics: highRiskChars
  };
}

function analyzeComparative(mergedData: any[]) {
  const sortedByScore = mergedData.sort((a, b) => b.credit_score - a.credit_score);
  const topPercentile = Math.floor(mergedData.length * 0.2);
  const bottomPercentile = Math.floor(mergedData.length * 0.2);
  
  const highScorers = sortedByScore.slice(0, topPercentile);
  const lowScorers = sortedByScore.slice(-bottomPercentile);

  const metrics = ['total_transactions', 'account_age_days', 'repayment_ratio', 'leverage_ratio'];
  const metricDifferences: Record<string, any> = {};

  metrics.forEach(metric => {
    const highAvg = mean(highScorers.map(h => h[metric] || 0));
    const lowAvg = mean(lowScorers.map(l => l[metric] || 0));
    
    metricDifferences[metric] = {
      high_scorers_avg: highAvg,
      low_scorers_avg: lowAvg,
      difference_ratio: lowAvg !== 0 ? highAvg / lowAvg : Infinity
    };
  });

  return {
    high_scorers: {
      count: highScorers.length,
      score_range: `${Math.min(...highScorers.map(h => h.credit_score))}-${Math.max(...highScorers.map(h => h.credit_score))}`
    },
    low_scorers: {
      count: lowScorers.length,
      score_range: `${Math.min(...lowScorers.map(l => l.credit_score))}-${Math.max(...lowScorers.map(l => l.credit_score))}`
    },
    metric_differences: metricDifferences
  };
}

function mean(values: number[]): number {
  return values.length > 0 ? values.reduce((sum, val) => sum + val, 0) / values.length : 0;
}

function median(values: number[]): number {
  if (values.length === 0) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 === 0 ? (sorted[mid - 1] + sorted[mid]) / 2 : sorted[mid];
}

function standardDeviation(values: number[]): number {
  if (values.length <= 1) return 0;
  const avg = mean(values);
  const squaredDiffs = values.map(val => Math.pow(val - avg, 2));
  return Math.sqrt(mean(squaredDiffs));
}

function calculateCorrelation(x: number[], y: number[]): number {
  if (x.length !== y.length || x.length === 0) return 0;
  
  const meanX = mean(x);
  const meanY = mean(y);
  
  let numerator = 0;
  let sumXSquared = 0;
  let sumYSquared = 0;
  
  for (let i = 0; i < x.length; i++) {
    const deltaX = x[i] - meanX;
    const deltaY = y[i] - meanY;
    
    numerator += deltaX * deltaY;
    sumXSquared += deltaX * deltaX;
    sumYSquared += deltaY * deltaY;
  }
  
  const denominator = Math.sqrt(sumXSquared * sumYSquared);
  return denominator === 0 ? 0 : numerator / denominator;
}

File: src/analyzer.py
"""
Analysis module for generating comprehensive reports on wallet scoring results.
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import logging
from typing import Dict, List, Any

class ScoreAnalyzer:
    """Analyzes wallet scoring results and generates comprehensive reports."""
    
    def __init__(self):
        self.logger = logging.getLogger(__name__)
        plt.style.use('seaborn-v0_8')
    
    def generate_comprehensive_analysis(self, scores_df: pd.DataFrame, features_df: pd.DataFrame) -> Dict[str, Any]:
        """
        Generate comprehensive analysis of scoring results.
        
        Args:
            scores_df: DataFrame with wallet scores
            features_df: DataFrame with engineered features
            
        Returns:
            Dictionary containing analysis results
        """
        self.logger.info("Generating comprehensive analysis...")
        
        analysis = {}
        
        # Score distribution analysis
        analysis['score_distribution'] = self._analyze_score_distribution(scores_df)
        
        # Behavioral pattern analysis
        analysis['behavioral_patterns'] = self._analyze_behavioral_patterns(scores_df, features_df)
        
        # Risk factor analysis
        analysis['risk_factors'] = self._analyze_risk_factors(scores_df, features_df)
        
        # Comparative analysis
        analysis['comparative_analysis'] = self._comparative_analysis(scores_df, features_df)
        
        return analysis
    
    def _analyze_score_distribution(self, scores_df: pd.DataFrame) -> Dict[str, Any]:
        """Analyze the distribution of credit scores."""
        scores = scores_df['credit_score']
        
        # Basic statistics
        distribution = {
            'total_wallets': len(scores),
            'mean_score': scores.mean(),
            'median_score': scores.median(),
            'std_score': scores.std(),
            'min_score': scores.min(),
            'max_score': scores.max(),
            'quartiles': {
                'q1': scores.quantile(0.25),
                'q2': scores.quantile(0.50),
                'q3': scores.quantile(0.75)
            }
        }
        
        # Score buckets (0-100, 100-200, etc.)
        bins = np.arange(0, 1001, 100)
        bucket_labels = [f"{bins[i]}-{bins[i+1]-1}" for i in range(len(bins)-1)]
        bucket_counts = pd.cut(scores, bins=bins, labels=bucket_labels, include_lowest=True).value_counts()
        
        distribution['bucket_distribution'] = {}
        for bucket, count in bucket_counts.items():
            percentage = (count / len(scores)) * 100
            distribution['bucket_distribution'][str(bucket)] = {
                'count': int(count),
                'percentage': round(percentage, 2)
            }
        
        # Risk category distribution
        category_counts = scores_df['risk_category'].value_counts()
        distribution['risk_category_distribution'] = {}
        for category, count in category_counts.items():
            percentage = (count / len(scores)) * 100
            distribution['risk_category_distribution'][category] = {
                'count': int(count),
                'percentage': round(percentage, 2)
            }
        
        return distribution
    
    def _analyze_behavioral_patterns(self, scores_df: pd.DataFrame, features_df: pd.DataFrame) -> Dict[str, Any]:
        """Analyze behavioral patterns across different score ranges."""
        # Merge scores with features
        merged_df = pd.merge(scores_df, features_df, on='wallet_address')
        
        patterns = {}
        
        # Define score groups
        score_groups = {
            'high_score': merged_df[merged_df['credit_score'] >= 700],
            'medium_score': merged_df[(merged_df['credit_score'] >= 400) & (merged_df['credit_score'] < 700)],
            'low_score': merged_df[merged_df['credit_score'] < 400]
        }
        
        # Analyze key behavioral metrics for each group
        behavioral_metrics = [
            'total_transactions', 'liquidation_ratio', 'repayment_ratio',
            'leverage_ratio', 'asset_diversity_score', 'account_age_days',
            'time_regularity_score', 'transaction_complexity'
        ]
        
        for group_name, group_data in score_groups.items():
            if len(group_data) > 0:
                patterns[group_name] = {}
                patterns[group_name]['count'] = len(group_data)
                
                for metric in behavioral_metrics:
                    if metric in group_data.columns:
                        patterns[group_name][metric] = {
                            'mean': float(group_data[metric].mean()),
                            'median': float(group_data[metric].median()),
                            'std': float(group_data[metric].std())
                        }
        
        return patterns
    
    def _analyze_risk_factors(self, scores_df: pd.DataFrame, features_df: pd.DataFrame) -> Dict[str, Any]:
        """Analyze key risk factors and their correlation with scores."""
        merged_df = pd.merge(scores_df, features_df, on='wallet_address')
        
        risk_factors = {}
        
        # Key risk indicators
        risk_indicators = [
            'liquidation_count', 'liquidation_ratio', 'leverage_ratio',
            'outstanding_debt', 'repayment_ratio', 'time_regularity_score',
            'amount_uniformity_score', 'gas_optimization_score'
        ]
        
        correlations = {}
        for indicator in risk_indicators:
            if indicator in merged_df.columns:
                correlation = merged_df['credit_score'].corr(merged_df[indicator])
                correlations[indicator] = float(correlation) if not np.isnan(correlation) else 0.0
        
        risk_factors['score_correlations'] = correlations
        
        # High-risk wallet analysis
        high_risk_wallets = merged_df[merged_df['credit_score'] < 400]
        if len(high_risk_wallets) > 0:
            risk_factors['high_risk_characteristics'] = {
                'count': len(high_risk_wallets),
                'avg_liquidations': float(high_risk_wallets['liquidation_count'].mean()) if 'liquidation_count' in high_risk_wallets.columns else 0,
                'avg_leverage': float(high_risk_wallets['leverage_ratio'].mean()) if 'leverage_ratio' in high_risk_wallets.columns else 0,
                'bot_like_ratio': float((high_risk_wallets['time_regularity_score'] > 2).mean()) if 'time_regularity_score' in high_risk_wallets.columns else 0
            }
        
        return risk_factors
    
    def _comparative_analysis(self, scores_df: pd.DataFrame, features_df: pd.DataFrame) -> Dict[str, Any]:
        """Compare high-scoring vs low-scoring wallets."""
        merged_df = pd.merge(scores_df, features_df, on='wallet_address')
        
        # Define comparison groups
        high_score_threshold = merged_df['credit_score'].quantile(0.8)
        low_score_threshold = merged_df['credit_score'].quantile(0.2)
        
        high_scorers = merged_df[merged_df['credit_score'] >= high_score_threshold]
        low_scorers = merged_df[merged_df['credit_score'] <= low_score_threshold]
        
        comparison = {
            'high_scorers': {
                'count': len(high_scorers),
                'score_range': f"{high_scorers['credit_score'].min():.0f}-{high_scorers['credit_score'].max():.0f}"
            },
            'low_scorers': {
                'count': len(low_scorers),
                'score_range': f"{low_scorers['credit_score'].min():.0f}-{low_scorers['credit_score'].max():.0f}"
            }
        }
        
        # Compare key metrics
        comparison_metrics = [
            'total_transactions', 'account_age_days', 'unique_assets',
            'repayment_ratio', 'liquidation_ratio', 'leverage_ratio'
        ]
        
        comparison['metric_differences'] = {}
        for metric in comparison_metrics:
            if metric in merged_df.columns:
                high_avg = high_scorers[metric].mean() if len(high_scorers) > 0 else 0
                low_avg = low_scorers[metric].mean() if len(low_scorers) > 0 else 0
                
                comparison['metric_differences'][metric] = {
                    'high_scorers_avg': float(high_avg),
                    'low_scorers_avg': float(low_avg),
                    'difference_ratio': high_avg / low_avg if low_avg != 0 else float('inf')
                }
        
        return comparison
    
    def generate_markdown_report(self, analysis: Dict[str, Any]) -> str:
        """Generate a markdown report from analysis results."""
        report = "# DeFi Credit Scoring Analysis Report\n\n"
        
        # Executive Summary
        report += "## Executive Summary\n\n"
        dist = analysis['score_distribution']
        report += f"This analysis covers **{dist['total_wallets']:,}** wallets with an average credit score of **{dist['mean_score']:.0f}**.\n\n"
        
        # Score Distribution
        report += "## Score Distribution\n\n"
        report += f"- **Mean Score**: {dist['mean_score']:.2f}\n"
        report += f"- **Median Score**: {dist['median_score']:.2f}\n"
        report += f"- **Standard Deviation**: {dist['std_score']:.2f}\n"
        report += f"- **Score Range**: {dist['min_score']:.0f} - {dist['max_score']:.0f}\n\n"
        
        # Bucket Distribution
        report += "### Score Distribution by Buckets\n\n"
        report += "| Score Range | Count | Percentage |\n"
        report += "|-------------|-------|------------|\n"
        
        for bucket, data in dist['bucket_distribution'].items():
            report += f"| {bucket} | {data['count']:,} | {data['percentage']:.1f}% |\n"
        
        report += "\n"
        
        # Risk Categories
        report += "### Risk Category Distribution\n\n"
        report += "| Risk Category | Count | Percentage |\n"
        report += "|---------------|-------|------------|\n"
        
        for category, data in dist['risk_category_distribution'].items():
            report += f"| {category} | {data['count']:,} | {data['percentage']:.1f}% |\n"
        
        report += "\n"
        
        # Behavioral Patterns
        if 'behavioral_patterns' in analysis:
            report += "## Behavioral Pattern Analysis\n\n"
            patterns = analysis['behavioral_patterns']
            
            for group_name, group_data in patterns.items():
                if isinstance(group_data, dict) and 'count' in group_data:
                    report += f"### {group_name.replace('_', ' ').title()} ({group_data['count']} wallets)\n\n"
                    
                    for metric, stats in group_data.items():
                        if metric != 'count' and isinstance(stats, dict):
                            report += f"- **{metric.replace('_', ' ').title()}**: {stats['mean']:.2f} (avg), {stats['median']:.2f} (median)\n"
                    
                    report += "\n"
        
        # Risk Factor Analysis
        if 'risk_factors' in analysis:
            report += "## Risk Factor Analysis\n\n"
            risk_factors = analysis['risk_factors']
            
            if 'score_correlations' in risk_factors:
                report += "### Risk Factor Correlations with Credit Score\n\n"
                report += "| Risk Factor | Correlation |\n"
                report += "|-------------|-------------|\n"
                
                sorted_correlations = sorted(risk_factors['score_correlations'].items(), 
                                           key=lambda x: abs(x[1]), reverse=True)
                
                for factor, correlation in sorted_correlations:
                    report += f"| {factor.replace('_', ' ').title()} | {correlation:.3f} |\n"
                
                report += "\n"
        
        # Comparative Analysis
        if 'comparative_analysis' in analysis:
            report += "## High vs Low Scorer Comparison\n\n"
            comp = analysis['comparative_analysis']
            
            report += f"- **High Scorers**: {comp['high_scorers']['count']} wallets (scores {comp['high_scorers']['score_range']})\n"
            report += f"- **Low Scorers**: {comp['low_scorers']['count']} wallets (scores {comp['low_scorers']['score_range']})\n\n"
            
            if 'metric_differences' in comp:
                report += "### Key Metric Differences\n\n"
                report += "| Metric | High Scorers | Low Scorers | Ratio |\n"
                report += "|--------|--------------|-------------|-------|\n"
                
                for metric, data in comp['metric_differences'].items():
                    ratio_str = f"{data['difference_ratio']:.2f}x" if data['difference_ratio'] != float('inf') else "∞"
                    report += f"| {metric.replace('_', ' ').title()} | {data['high_scorers_avg']:.2f} | {data['low_scorers_avg']:.2f} | {ratio_str} |\n"
        
        report += "\n---\n\n*Report generated by DeFi Credit Scoring System*\n"
        
        return report